### PR TITLE
Better name parsing

### LINKED ISSUE #11: Better parsing of image names
> Probably the most comprehensive logic to parse image names is here: https://github.com/projectatomic/atomic-reactor/blob/master/atomic_reactor/util.py#L57

### PR DESCRIPTION
**Pull request by lachmanfrantisek**
Resolves #11 

- Added also support for instance of conu Image/Container class in target (needs some parameter renames)
- Image id is not supported now. (Need some work on the conu side.)

### CODE REVIEW

File: colin/core/target.py [hunk 1]
```diff
@@ -92,3 +100,58 @@ def is_compatible(target_type, check_class, severity, tags):
     return (target_type == TargetType.DOCKERFILE and isinstance(check_class, DockerfileCheck)) \
            or (target_type == TargetType.CONTAINER and isinstance(check_class, ContainerCheck)) \
            or (target_type == TargetType.CONTAINER_IMAGE and isinstance(check_class, ImageCheck))
+
+
+class ImageName(object):
```
Comment (TomasTomecek) [inline colin/core/target.py]: Would be awesome to have this in conu. At some point.

File: colin/core/target.py [hunk 2]
```diff
@@ -92,3 +100,58 @@ def is_compatible(target_type, check_class, severity, tags):
...
+            else:
+                result.namespace = s[0]
+        elif len(s) == 3:
+            result.registry = s[0]
+            result.namespace = s[1]
+        result.repository = s[-1]
+
+        for sep in '@:':
+            try:
+                result.repository, result.tag = result.repository.rsplit(sep, 1)
```
Comment (TomasTomecek) [inline colin/core/target.py]: this will be confusing for digest, since digest is not tag

File: colin/core/target.py (hunk 1 above)
Comment (TomasTomecek) [inline colin/core/target.py]: This class needs tests badly.

File: colin/core/target.py [hunk 3]
```diff
@@ -92,3 +100,58 @@ def is_compatible(target_type, check_class, severity, tags):
...
+        return result
+
+    def __str__(self):
+        return "Image: registry='{}' namespace='{}' repository='{}' tag='{}'".format(self.registry,
+                                                                                     self.namespace,
+                                                                                     self.repository,
+                                                                                     self.tag)
+
+    @property
+    def name(self):
```
Comment (TomasTomecek) [inline colin/core/target.py]: docs please, what's a name? docker terminology is so confusing

File: colin/core/target.py (hunk 1 above)
Comment (lachmanfrantisek) [reply colin/core/target.py]: Yes, I will look at it with [this issue](https://github.com/fedora-modularity/conu/issues/185).

File: colin/core/target.py [hunk 4]
```diff
@@ -32,44 +32,51 @@
...
-                name_split = target_name.split(':')
-                if len(name_split) == 2:
-                    name, tag = name_split
-                    image = backend.ImageClass(repository=name,
-                                               tag=tag,
+
+                image_name = ImageName.parse(target)
+                logger.debug(str(image_name))
+
+                logger.debug("Finding image '{}' with tag '{}'.".format(image_name.name, image_name.tag))
```
Comment (jpopelka) [inline colin/core/target.py]: Do we need both debug messages? The above one looks like a leftover to me.

File: colin/core/target.py (hunk 1 above)
Comment (lachmanfrantisek) [reply colin/core/target.py]: Done.

Do you have some other ugly image names?

File: colin/core/target.py (hunk 3 above)
Comment (lachmanfrantisek) [reply colin/core/target.py]: Done.

File: colin/core/target.py (hunk 2 above)
Comment (lachmanfrantisek) [reply colin/core/target.py]: I've added also `digest` property.

File: colin/core/target.py (hunk 4 above)
Comment (lachmanfrantisek) [reply colin/core/target.py]: Good point. Thanks.

File: tests/test_image_name.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    ("fedora", "Image: registry='None' namespace='None' repository='fedora' tag='None' digest='None'"),
+    ("fedora:27", "Image: registry='None' namespace='None' repository='fedora' tag='27' digest='None'"),
+    ("docker.io/fedora", "Image: registry='docker.io' namespace='None' repository='fedora' tag='None' digest='None'"),
+    ("docker.io/fedora:latest",
+     "Image: registry='docker.io' namespace='None' repository='fedora' tag='latest' digest='None'"),
+    ("docker.io/modularitycontainers/conu",
+     "Image: registry='docker.io' namespace='modularitycontainers' repository='conu' tag='None' digest='None'"),
+    ("docker.io/centos/postgresql-96-centos7",
+     "Image: registry='docker.io' namespace='centos' repository='postgresql-96-centos7' tag='None' digest='None'"),
+    ("brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel6",
```
Comment (TomasTomecek) [inline tests/test_image_name.py]: please don't use internal hostnames; `registry.example.com:8888` should do just fine

File: tests/test_image_name.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+     "Image: registry='brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888' namespace='None' repository='rhel6' tag='None' digest='None'"),
+    ("brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel6:guest-rhel-6.10-docker-26365-20180322014912",
+     "Image: registry='brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888' namespace='None' repository='rhel6' tag='guest-rhel-6.10-docker-26365-20180322014912' digest='None'"),
+    ("fedora@sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
+     "Image: registry='None' namespace='None' repository='fedora' tag='None' digest='sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'"),
+])
+def test_image_class(string_input, image_str):
+    image_name = ImageName.parse(string_input)
+    assert str(image_name) == image_str
```
Comment (TomasTomecek) [inline tests/test_image_name.py]: this seems fragile to me; would it make sense to do `assert image_name == expected_image` and have the parametrization set as `(image_str, ImageName(registry=..., namespace=...)`?

File: tests/test_image_name.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+     "Image: registry='docker.io' namespace='None' repository='fedora' tag='latest' digest='None'"),
+    ("docker.io/modularitycontainers/conu",
+     "Image: registry='docker.io' namespace='modularitycontainers' repository='conu' tag='None' digest='None'"),
+    ("docker.io/centos/postgresql-96-centos7",
+     "Image: registry='docker.io' namespace='centos' repository='postgresql-96-centos7' tag='None' digest='None'"),
+    ("brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel6",
+     "Image: registry='brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888' namespace='None' repository='rhel6' tag='None' digest='None'"),
+    ("brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhel6:guest-rhel-6.10-docker-26365-20180322014912",
+     "Image: registry='brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888' namespace='None' repository='rhel6' tag='guest-rhel-6.10-docker-26365-20180322014912' digest='None'"),
+    ("fedora@sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
```
Comment (TomasTomecek) [inline tests/test_image_name.py]: I would add another one which also has registry + name + digest.

File: colin/core/target.py [hunk 8]
```diff
@@ -32,44 +32,50 @@
...
+        :param target: str or instance of Image/Container
         :return: Container/Image
         """
-        logger.debug("Finding target '{}'.".format(target_name))
+        logger.debug("Finding target '{}'.".format(target))
+
+        if isinstance(target, Image):
+            return target
+        if isinstance(target, Container):
+            return target
```
[the remark this answers is not in the document]
Comment (lachmanfrantisek) [reply colin/core/target.py]: Thanks, don't know that.

**Comment by lachmanfrantisek** [comment]
@TomasTomecek  I've added tests for the `name` property of `ImageName` class.


Also, I've correct the CLI (Do I have to move this to another pull request?):

```diff
-def cli(target, config, config_file, debug, json, stat, print_checks, verbose, version):
+def cli(target, config, config_file, debug, json, stat, print_checks, verbose):
```

**Comment by TomasTomecek** [comment]
Tests are passing, lgtm, nicely done!

**Comment by lachmanfrantisek** [comment]
@jpopelka @TomasTomecek  Thank you for the review.

**Pull request merged**